modules/stochastic_tools/src/utils/CovarianceFunction.C

```diff
@@ -0,0 +1,402 @@
+}
+
+std::unique_ptr<CovarianceKernel>
+makeCovarianceKernel(const MooseEnum & kernel_type, const GaussianProcessTrainer * gp)
+{
+
+  std::vector<Real> length_factor = gp->getParam<std::vector<Real>>("length_factor");
+  Real signal_variance = gp->getParam<Real>("signal_variance");
+  Real noise_variance = gp->getParam<Real>("noise_variance");
+
+  if (kernel_type == 0)
+  {
+    std::unique_ptr<CovarianceKernel> ptr = libmesh_make_unique<SquaredExponential>();
+    ptr->set_signal_variance(signal_variance);
+    ptr->set_noise_variance(noise_variance);
+    ptr->set_length_factor(length_factor);
+    return ptr;
+  }
+
+  if (kernel_type == 1)
+  {
+    std::unique_ptr<CovarianceKernel> ptr = libmesh_make_unique<Exponential>();
+    Real gamma = gp->getParam<Real>("gamma");
+    ptr->set_signal_variance(signal_variance);
+    ptr->set_noise_variance(noise_variance);
+    ptr->set_length_factor(length_factor);
+    ptr->set_gamma(gamma);
+    return ptr;
+  }
+
+  if (kernel_type == 2)
+  {
+    ::mooseError("General Matern covariance function not yet implemented. Half-integer Matern "
+                 "implemented in MaternHalfInt if desired.");
+    return nullptr;
+  }
+
+  if (kernel_type == 3)
+  {
+    std::unique_ptr<CovarianceKernel> ptr = libmesh_make_unique<MaternHalfInt>();
+    unsigned int p = gp->getParam<unsigned int>("p");
+    ptr->set_signal_variance(signal_variance);
+    ptr->set_noise_variance(noise_variance);
+    ptr->set_length_factor(length_factor);
+    ptr->set_p(p);
+    return ptr;
+  }
+
+  ::mooseError("Unknown covariance function type");
+  return nullptr;
+}
+
+void
+CovarianceKernel::store(std::ostream & /*stream*/, void * /*context*/) const
+{
+  // Cannot be pure virtual because for dataLoad operations the base class must be constructed
+  ::mooseError("Covariance Function child class must override 'store()' method.");
+}
+
+RealEigenMatrix
+CovarianceKernel::compute_K(const RealEigenMatrix /*x*/,
+                            const RealEigenMatrix /*xp*/,
+                            const bool /*is_self_covariance*/) const
+{
+  ::mooseError("Covariance function has not been implemented.");
+  RealEigenMatrix tmp(1, 1);
+  return tmp;
+}
+
+void
+CovarianceKernel::set_signal_variance(const Real sigma_f_squared)
+{
+  _sigma_f_squared = sigma_f_squared;
+}
+
+void
+CovarianceKernel::set_noise_variance(const Real sigma_n_squared)
+{
+  _sigma_n_squared = sigma_n_squared;
+}
+
+void
+CovarianceKernel::set_length_factor(const std::vector<Real> length_factor)
+{
+  _length_factor = length_factor;
+}
+
+void
+CovarianceKernel::set_gamma(const Real /*gamma*/)
+{
+  ::mooseError(
+      "Attempting to set gamma hyper parameter for covariance function which does not utilize.");
+}
+
+void
+CovarianceKernel::set_p(const unsigned int /*p*/)
+{
+  ::mooseError(
+      "Attempting to set p hyper parameter for covariance function which does not utilize.");
+}
```

Comment:
I think all these functions can be replaced by constructor arguments.

Reply:
See other reply, these will be needed in near future development

Reply:
Done, see CovarianceFunction MOOSE Objects